runtime/lua/vim/_comment.lua

```diff
@@ -0,0 +1,252 @@
+local function make_comment_parts(ref_position)
+  local cs = get_commentstring(ref_position)
+
+  if cs == nil or cs == '' then
+    vim.api.nvim_echo({ { "Option 'commentstring' is empty.", 'WarningMsg' } }, true, {})
+    return { left = '', right = '' }
+  end
+
+  if not (type(cs) == 'string' and string.find(cs, '%%s') ~= nil) then
+    error(vim.inspect(cs) .. " is not a valid 'commentstring'.")
+  end
+
+  -- Structure of 'commentstring': <left part> <%s> <right part>
+  local left, right = string.match(cs, '^(.-)%%s(.-)$')
+  return { left = left, right = right }
+end
+
+---@param parts vim._comment.Parts
+---@return fun(line: string): boolean
+local function make_comment_check(parts)
+  local l_esc, r_esc = vim.pesc(parts.left), vim.pesc(parts.right)
+
+  -- Commented line has the following structure:
+  -- <possible whitespace> <left> <anything> <right> <possible whitespace>
+  local nonblank_regex = '^%s-' .. l_esc .. '.*' .. r_esc .. '%s-$'
+
+  -- Commented blank line can have any amoung of whitespace around parts
+  local blank_regex = '^%s-' .. vim.trim(l_esc) .. '%s*' .. vim.trim(r_esc) .. '%s-$'
+
+  return function(line)
+    return string.find(line, nonblank_regex) ~= nil or string.find(line, blank_regex) ~= nil
+  end
+end
+
+---@param lines string[]
+---@param parts vim._comment.Parts
+---@return string indent
+---@return boolean is_commented
+local function get_lines_info(lines, parts)
+  local comment_check = make_comment_check(parts)
+
+  local is_commented = true
+  local indent_width, indent_width_cur = math.huge, math.huge
+  ---@type integer
+  local indent
+  ---@type integer
+  local indent_cur
+
+  for _, l in ipairs(lines) do
+    -- Update lines indent: minimum of all indents except blank lines
+    ---@diagnostic disable-next-line:cast-local-type
+    _, indent_width_cur, indent_cur = string.find(l, '^(%s*)')
+    local is_blank = indent_width_cur == string.len(l)
+
+    -- NOTE: Copying actual indent instead of recreating it with `indent_width`
+    -- allows to handle both tabs and spaces
+    if indent_width_cur < indent_width and not is_blank then
+      ---@diagnostic disable-next-line:cast-local-type
+      indent_width, indent = indent_width_cur, indent_cur
+    end
+
+    -- Update comment info: lines are commented if every line is commented
+    if is_commented then
+      is_commented = comment_check(l)
+    end
+  end
+
+  -- `indent` can still be `nil` in case all `lines` are empty
+  return indent or '', is_commented
+end
+
+---@param x string
+---@return boolean is_blank
+local function is_blank(x)
+  return string.find(x, '^%s*$') ~= nil
+end
+
+---@param parts vim._comment.Parts
+---@param indent string
+---@return fun(line: string): string
+local function make_comment_function(parts, indent)
+  local prefix, nonindent_start, suffix = indent .. parts.left, string.len(indent) + 1, parts.right
+  local blank_comment = indent .. vim.trim(parts.left) .. vim.trim(parts.right)
```

Comment:
Can this just go inside the returned closure?

Reply:
One of the points to returning closure is so that those values are not evaluated on every call to comment or (later) uncomment function. I think having them outside is better here.